### DISCUSSION TITLE
Carbon parsing error on datetime cast when binding directly to model properties

### DISCUSSION THREAD

**Discussion by richbkuk** [livewire/livewire#3143] [7. Bugs (archived)]
### Description
When a model has an attribute cast as datetime and you're binding directly to model properties in Livewire, if the user submits a value that Carbon cannot parse, Carbon returns an error when ideally, validation should have picked up it up first before the value is applied to the model.
I believe the issue is that the value submitted is being applied to the model before validation takes place?

### Exact steps to reproduce
- Create a model with an attribute cast as datetime
- 'Make' the empty model in the mount() Livewire method
- Add a validation entry for it, as per the docs
- From the view (front end form), submit an invalid date format such as 'non-date'
- Receive error: Could not parse 'non-date': DateTime::__construct(): Failed to parse time string (non-test) at position 0 (n): The timezone could not be found in the database

### Stripped-down, copy-pastable code snippets
Model:
```php
class Event extends Model
{
    protected $casts = [
        'start' => 'datetime'
    ];
}
```

Livewire Component:
```php
class CreateEvent extends Component
{
    public $event;
	
    protected $rules = [
        'event.start' => 'required|date',
    ];
	
    public function mount() {
        $this->event = Event::make();
    }
	
    public function submit() {
        $this->validate();

        $this->event->save();
    }
	
    public function render() {
        return view('livewire.create-event');
    }
}
```

Livewire View:
```html
<form wire:submit.prevent="submit">
    <input type="text" wire:model.defer="event.start" />
    <button type="submit">Submit</button>
</form>
```

**Context**
- Livewire version: 2.2.5
- Laravel version: 8.6.0
- Browser: Chrome 85.0.4183.121
- PHP: 7.3

**Comment by uteq** [comment]
@richbkuk What I did to resolve this is added the start date to the models $dates property and removed it from the $casts. This prevented throwing the error. This is more a workaround I guess than an actual solution. Hopefully this helps someone.

**Comment by JuanRangel** [comment]
After running into the same issue I did some digging and noticed that when trying to set my date attribute with Livewire it was never reaching my livewire method. The error seems to happen in the HasAttribute trait in Laravel. 

Even if you remove the date validation the error still exists

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by zepfietje** [comment]
REOPEN

**Comment by mikebronner** [comment]
Having this issue as well.

**Comment by mdurao** [comment]
Is there a solution for this?

**Comment by AdamEsterle** [comment]
I am having this issue as well. My discoveries:

SUMMARY
`$this-validate()`
in livewire will FIRST CAST the value, then check validation
Then when you run
`$this->model->save()`
That will try to save the USER INPUTED DATA FROM THE REQUEST. This is probably a security concern.
Basically, it validates a casted set of data, then tries to save a pre-casted set of data.

SPECIFICS
It is because in the Laravel Validator, it is doing Arr::get() which will FIRST cast the model's value BEFORE validating. (when an Eloquent model gets converted to an array, it casts its values)

There is also a problem in that Livewire converts models to arrays (which casts their values) before showing them in the UI. This is easily seen:
1. Make a model with a property that casts to integer
2. Put in livewire class (public User $user)
3. wire:model="user.field" on an input
4. watch the input autoconvert (you cannot type 10.5, it'll update to 10)

**Comment by AdamEsterle** [comment]
See my discussion here for more details
https://github.com/livewire/livewire/discussions/3400

**Reply by ChrisSantiago82** [reply]
I always had problems when I wanted to bind a carbon casted field to a date input.

Since a while I'm using setter and getter in the models for that. Like this it's still possible to use the carbon functionality where needed, but it's possible to bind a date directly to an input in livewire

**Reply by stefanocurnis** [reply]
This is still a problem after years. It's really a pain to use Livewire without this. There's even an unmerged pull request that fixes the issue.

**Reply by joshhanley** [reply]
@stefanocurnis see my comment here for how cast values before the update has finished https://github.com/livewire/livewire/discussions/3370#discussioncomment-1023847

Hope this helps!
